Dialog.js

```diff
@@ -437,14 +437,33 @@ define([
 					viewport.h *= this.maxRatio;
 
 					var bb = domGeometry.position(this.domNode);
-					if(bb.w >= viewport.w || bb.h >= viewport.h){
+					this._shrunk = false;
+					// First check and limit width, because limiting the width may increase the height due to word wrapping.
+					if(bb.w >= viewport.w){
 						dim = {
-							w: Math.min(bb.w, viewport.w),
-							h: Math.min(bb.h, viewport.h)
+							w: viewport.w
 						};
+						domGeometry.setMarginBox(this.domNode, dim);
+						bb = domGeometry.position(this.domNode);
 						this._shrunk = true;
-					}else{
-						this._shrunk = false;
+					}
+					// Now check and limit the height
+					if(bb.h >= viewport.h){
+						if(!dim){
+							dim = {
+								w: bb.w
```

Comment:
It seems unnecessary to set the width again; if the width needed to be adjusted then didn't you already do it on line 446?

More importantly, if line 444 reduced the width, won't this line revert that change, making the Dialog too wide for the viewport?

Reply:
I somehow assumed that `domGeometry.setMarginBox` always needs a width and height. If this is not the case, the code can be simplified.
I think the second call should not revert the width limitation, because `bb` is read again in line 447.

Reply:
Yah, looks like setMarginBox() calls setBox(), which only tries to set the width if it's specified.

I didn't notice the new position() call on line 446... I wonder if all the position calls should be replaced by references to offsetWidth or offsetHeight.